src/v0/util/index.js

```diff
@@ -2296,14 +2297,45 @@ const applyJSONStringTemplate = (message, template) =>
   }).evaluate(message);
 
 /**
- * This groups the events by destination ID and source ID.
+ * This groups the events by destination ID, source ID, and optionally by destination config.
  * Note: sourceID is only used for rETL events.
+ *
+ * The function checks the hasDynamicConfig flag on the destination to determine if events
+ * should be grouped by destination config as well. This is important for destinations that
+ * use dynamic configuration where different events might need different config values.
+ *
+ * For backward compatibility with older server versions where the hasDynamicConfig flag
+ * might not be available, the function treats undefined flags as if dynamic config might
+ * be present (only skips grouping by config when the flag is explicitly false).
+ *
  * @param {*} events The events to be grouped.
  * @returns {array} The array of grouped events.
  */
 const groupRouterTransformEvents = (events) =>
   Object.values(
-    lodash.groupBy(events, (ev) => [ev.destination?.ID, ev.context?.sources?.job_id || 'default']),
+    lodash.groupBy(events, (ev) => {
+      // Check if hasDynamicConfig flag is available
+      // If undefined (older server versions), process all events as if they might have dynamic config
+      // Only skip processing if the flag is explicitly false
+      const hasDynamicConfigFlag = ev.destination?.hasDynamicConfig;
```

Comment:
do we have environment variable to control this new behaviour ? either in rudder-server while generating `hasDynamicConfig` flag or other-wise should have it here